src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/StringSearchValues.cs

```diff
@@ -0,0 +1,391 @@
+            }
+
+            var uniqueValues = new HashSet<string>(values.Length, ignoreCase ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal);
+
+            foreach (string value in values)
+            {
+                ArgumentNullException.ThrowIfNull(value, nameof(values));
+
+                uniqueValues.Add(value);
+            }
+
+            if (uniqueValues.Contains(string.Empty))
+            {
+                return new SingleStringSearchValuesFallback<SearchValues.FalseConst>(string.Empty, uniqueValues);
+            }
+
+            Span<string> normalizedValues = new string[uniqueValues.Count];
+            int i = 0;
+            foreach (string value in uniqueValues)
+            {
+                normalizedValues[i++] = NormalizeIfNeeded(value, ignoreCase);
+            }
+            Debug.Assert(i == normalizedValues.Length);
+
+            if (normalizedValues.Length == 1)
+            {
+                // Avoid the overhead of building the AhoCorasick trie for single-value inputs.
+                AnalyzeValues(normalizedValues, ref ignoreCase, out bool ascii, out bool asciiLettersOnly, out _, out _);
+                return CreateForSingleValue(normalizedValues[0], uniqueValues, ignoreCase, ascii, asciiLettersOnly);
+            }
+
+            // Aho-Corasick's ctor expects values to be sorted by length.
+            normalizedValues.Sort(static (a, b) => a.Length.CompareTo(b.Length));
+
+            // We may not end up choosing Aho-Corasick as the implementation, but it has a nice property of
+            // finding all the unreachable values during the construction stage, so we build the trie early.
+            HashSet<string>? unreachableValues = null;
+            var ahoCorasickBuilder = new AhoCorasickBuilder(normalizedValues, ignoreCase, ref unreachableValues);
+
+            if (unreachableValues is not null)
+            {
+                // Some values are exact prefixes of other values.
+                // Exclude those values now to reduce the number of buckets and make verification steps cheaper during searching.
+                normalizedValues = RemoveUnreachableValues(normalizedValues, unreachableValues);
+            }
+
+            SearchValues<string> searchValues = CreateFromNormalizedValues(normalizedValues, uniqueValues, ignoreCase, ref ahoCorasickBuilder);
+            ahoCorasickBuilder.Dispose();
+            return searchValues;
+
+            static string NormalizeIfNeeded(string value, bool ignoreCase)
+            {
+                if (ignoreCase && (value.AsSpan().ContainsAnyInRange('a', 'z') || !Ascii.IsValid(value)))
+                {
+                    string upperCase = string.FastAllocateString(value.Length);
+                    int charsWritten = Ordinal.ToUpperOrdinal(value, new Span<char>(ref upperCase.GetRawStringData(), upperCase.Length));
+                    Debug.Assert(charsWritten == upperCase.Length);
+                    value = upperCase;
```

Comment:
Should we factor this out into an internal string.ToUpperOrdinal?

Reply:
This is the only use I see that actually needs a new string (as opposed to filling a scratch buffer), but I can move it to a separate helper.